_posts/2020/2020-02-11-unit-testing-in-react.md

```diff
@@ -0,0 +1,251 @@
+to assert whether the `ProductList` is rendering them correctly. This way our tests becomes a lot more predictable.
+
+However... Can you do it? It's now difficult to mock `fetchProducts`, because you'd have to override the `ProductList.jsx` file.
+There comes another hint: try to inject your external dependencies such that you can mock them! You could pass `fetchProducts` as a prop instead:
+
+```javascript
+const defaultFetchProducts = (page) => fetch(`http://localhost/products?page=${page}`)
+  .then(response => response.json())
+  .catch((error) => console.log(error))
+
+const ProductList = ({ fetchProducts=defaultFetchProducts }) => {
+  const [products, setProducts] = useState([])
+  const [page, setPage] = useState(1)
+  const loadProducts = async () => {
+    const response = await fetchProducts(page)
+    if(response) {
+      const newProducts = response.products
+      setProducts(products.concat(newProducts))
+      setPage(page + 1)
+    }
+  }
+  return <div>
+    <button onClick={loadProducts}>
+       Load more!
+    </button>
+    {products.map(product => <Product {...product} />)}
+  </div>
+}
+```
+
+And what about the test?
+
+```javascript
+import React from 'react'
+import { fireEvent, render } from '@testing-library/react'
+import ProductList from '../ProductList'
+import Bluebird from 'bluebird'
+import { act } from 'react-dom/test-utils'
+
+it('should render a button to load products', async () => {
+  const fetchProducts = jest.fn().mockImplementation((page) => Promise.resolve({
+    products: [{
+      id: `dummy-id-${page}`,
+      name: 'dummy-product',
+      price: '10 £',
+    }]
+  }))
+
+  const { getByText } = render(<ProductList fetchProducts={fetchProducts}/>)
+
+  const loadMoreButton = getByText('Load more!')
+
+  // click the first time
+  fireEvent.click(loadMoreButton)
+  await act(() => Bluebird.delay(500))
+
+  // expect one product
+  expect(getByText('dummy-id-1')).toBeTruthy()
+  expect(fetchProducts).toHaveBeenCalledTimes(1)
+  expect(fetchProducts).toHaveBeenCalledWith(1)
+
+  //click the second time
+  fireEvent.click(loadMoreButton)
+  await act(() => Bluebird.delay(500))
+
+  // expect two products
+  expect(getByText('dummy-id-1')).toBeTruthy()
+  expect(fetchProducts).toHaveBeenCalledTimes(2)
+  expect(fetchProducts).toHaveBeenCalledWith(2)
+})
+```
+
+This test can already be taken seriously! It checks whether we called `fetchProducts` with the right arguments and whether we
+use the result of these calls in a meaningful way.
+
+On the repo, this approach is on branch `mocking-fetch-products`.
+
+## Taking it to the next level
+
+The former approach has, nevertheless, its drawbacks. We rely completely on the mock of `fetchProducts`. How can we know if it would
+hit the right API endpoints?
+There's were the awesome `nock` library comes along. You can also mock the HTTP requests! Calling `nock('http://localhost')` will
```

Comment:
Same as written in the other annotation: I think we shouldn't use the package names, but instead the official spelling of the names according to the corresponding libraries (and possibly link them for the first occurrence in the text).

```suggestion
That's were the awesome Nock library comes along. You can also mock the HTTP requests! Calling `nock('http://localhost')` will
```

Reply:
done - but here I really mean "There"